### PR TITLE
NG-858 Toggle speakerphone whenever the selected audio device is changed

### PR DESCRIPTION
**Pull request by f1shear**

### CODE REVIEW

**Review by euclides-jr — APPROVED** [review]

**Review by marton78 — APPROVED** [review]

**Review by javipascual — APPROVED** [review]

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java [hunk 1]
```diff
@@ -443,6 +443,8 @@ private void selectSpeakerAsAudioOutput() {
     private void connectToRoom(String roomName, String accessToken) {
         audioSwitch = new AudioSwitch(getApplicationContext());
 
+        selectSpeakerAsAudioOutput();
```
Comment (javipascual) [inline android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: why this? what if the user is on headphones?

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java [hunk 2]
```diff
@@ -323,9 +327,13 @@ private void createAudioAndVideoTracks() {
...
         localVideoTrack = LocalVideoTrack.create(this, true, cameraCapturerCompat.getVideoCapturer(), LOCAL_VIDEO_TRACK_NAME);
 
+
         audioSwitch.start(
-            (audioDevices, audioDevice) -> {
-                // TODO Enable user select audio device
+            (audioDevices, selectedAudioDevice) -> {
+                if(selectedAudioDevice != null) {
+                    if (selectedAudioDevice instanceof AudioDevice.Earpiece) selectSpeakerAsAudioOutput();
+                    else audioSwitch.selectDevice(selectedAudioDevice);
```
Comment (javipascual) [inline android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: is this necessary?

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java [hunk 3]
```diff
@@ -423,6 +431,15 @@ private void removeParticipantVideo(VideoTrack videoTrack) {
         videoTrack.removeRenderer(primaryVideoView);
     }
 
+
+    private void selectSpeakerAsAudioOutput() {
+
+        List<AudioDevice> devices = audioSwitch.getAvailableAudioDevices();
+        for (AudioDevice audioDevice: devices) {
+            if(audioDevice instanceof AudioDevice.Speakerphone) audioSwitch.selectDevice(audioDevice);
```
Comment (javipascual) [inline android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: return, or break

**Review by javipascual — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java (hunk 2 above)
Comment (euclides-jr) [reply android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: The else is not necessary

**Review by euclides-jr — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java (hunk 2 above)
Comment (javipascual) [reply android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: I think so too

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java [hunk 4]
```diff
@@ -323,9 +327,12 @@ private void createAudioAndVideoTracks() {
         cameraCapturerCompat = new CameraCapturerCompat(this, getAvailableCameraSource());
         localVideoTrack = LocalVideoTrack.create(this, true, cameraCapturerCompat.getVideoCapturer(), LOCAL_VIDEO_TRACK_NAME);
 
+
         audioSwitch.start(
-            (audioDevices, audioDevice) -> {
-                // TODO Enable user select audio device
+            (audioDevices, selectedAudioDevice) -> {
+                if(selectedAudioDevice != null) {
+                    if (selectedAudioDevice instanceof AudioDevice.Earpiece) selectSpeakerAsAudioOutput();
```
Comment (javipascual) [inline android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: why not just one `if ( && )` ?

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java [hunk 5]
```diff
@@ -423,9 +430,23 @@ private void removeParticipantVideo(VideoTrack videoTrack) {
...
+                break;
+            }
+        }
+    }
+
     private void connectToRoom(String roomName, String accessToken) {
         audioSwitch = new AudioSwitch(getApplicationContext());
 
+        AudioDevice selectedDevice = audioSwitch.getSelectedAudioDevice();
+        if (selectedDevice instanceof  AudioDevice.Earpiece) selectSpeakerAsAudioOutput();
```
Comment (javipascual) [inline android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: one space too much :D

**Review by javipascual — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java (hunk 1 above)
Comment (f1shear) [reply android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: now, there is a condition to check if its earpiece

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java (hunk 2 above)
Comment (f1shear) [reply android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: yes, you are right.  its not needed I thought parameter `selectedAudioDevice` was something that the default active output device

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java (hunk 3 above)
Comment (f1shear) [reply android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: done

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java (hunk 5 above)
Comment (f1shear) [reply android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: I think the space here makes sense, the above two lines belong to difference concept than below

**Review by euclides-jr — APPROVED** [review]

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java (hunk 5 above)
Comment (javipascual) [reply android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: space, no line break, between `instanceof` and `AudioDevice.Earpiece`

Comment (f1shear) [reply android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: oh, sorry I misunderstood

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java (hunk 4 above)
Comment (f1shear) [reply android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: changed.

File: android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java (hunk 5 above)
Comment (f1shear) [reply android/src/main/java/com/media4care/twilio/video/TwilioVideoActivity.java]: fixed

**Review by javipascual — APPROVED** [review]

**Pull request merged**